---
title: "VCF data"
author: "Brian J. Knaus"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{vcf data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Most variant calling pipelines result in files containing variant information.
The [variant call format (vcf)](http://samtools.github.io/hts-specs/ "VCF format at hts-specs") is an increasingly popular format for this data.
The genotypes in these files are typically intended to be filtered as an attempt to remove false positives or other problematic variants.
A first step in working with this data is to understand their contents.


## Three sections


A VCF file can be thought of as having three sections: a **meta region**, a **fix region** and a **gt region**.
The meta region is at the top of the file.
The information in the meta region defines the abbreviations used elsewhere in the file.
It may also document software used to create the file as well as parameters used by this software.
Below the meta region, the data are tabular.
The first eight columns of this table contain information about each variant.
This data may be common over all variants, such as its chromosomal position, or a summary over all samples, such as quality metrics.
These data are fixed, or the same, over all samples.
Beginning at column ten is a column for every sample.
The values in these columns are information for each sample and each variant.
The organization of each cell containing a genotype and associated information is specified in column nine.
The location of these three regions within a file can be represented by the cartoon below.


```{r, fig.cap="Cartoon representation of VCF file organization", echo=FALSE, fig.height=4, fig.width=4, fig.align='center', }
par(mar=c(0.1,0.1,0.1,0.1))
plot(c(0,5), c(0,5), type="n", frame.plot=FALSE, axes=FALSE, xlab="", ylab="")
rect(xleft=0, ybottom=4, xright=3, ytop=5)
rect(xleft=0, ybottom=0, xright=2, ytop=4)
rect(xleft=2, ybottom=0, xright=5, ytop=4)
text(1.5, 4.5, "meta")
text(1.0, 2.5, "fix")
text(3.5, 2.5, "gt")
par(mar=c(5,4,4,2))
```


The VCF file definition is flexible.
This means that there are slots for certain types of data, but any particular software which creates a VCF file does not necessarily use them all.
Similarly, authors have the opportunity to include new forms of data, forms which may not have been foreseen by the authors of the VCF definition.
The result is that all VCF files do not contain the same information.


For this vignette, we'll use the example data provided with vcfR.


```{r}
library(vcfR)
data(vcfR_example)
```


## The meta region


The meta region contains information about the file and its creation, as well as information to interpret abbreviations used elsewhere in the file.
Each line of the meta region begins with a double pound sign ('##').
The example which comes with vcfR is shown below.
(Only the first 10 lines are shown for brevity.)



```{r, echo=FALSE}
strwrap(vcf@meta[1:7])
```


The first line contains the version of the VCF format used in the file.
This line is required.
The second line specifies the software which created the VCF file.
This is not required, so not all VCF files include it.
When they do, the file becomes self documenting.
Note that the alignment software is not included here because it was used upstream of the VCF file's creation (aligners typically create \*.SAM or \*.BAM format files).
Because the file can only include information about the software that created it, the entire pipeline does not get documented.
Some VCF files may contain a line for every chromosome (or supercontig or contig depending on your genome), so they may become rather long.
Here, the remaining lines contain INFO and FORMAT specifications which define abbreviations used in the fix and gt portions of the file.



## The fix region

The fix region contains information for each variant which is sometimes summarized over all samples.
The first eight columns of the fixed region and are titled CHROM, POS, ID, REF, ALT, QUAL, FILTER and INFO.
This is per variant information which is 'fixed', or the same, over all samples.
The first two columns indicate the location of the variant by chromosome and position within that chromosome.
Here, the ID field has not been used, so it consists of missing data (NA).
The REF and ALT columns indicate the reference and alternate allelic states.
When multiple alternate allelic states are present they are delimited with commas.
The QUAL column attempts to summarize the quality of each variant over all samples.
The FILTER field is not used here but could contain information on whether a variant has passed some form of quality assessment. 


```{r, echo=FALSE}
vcf@fix[1:6, 1:7]
```



The eigth column, titled INFO, is a semicolon delimited list of information.
It can be rather long and cumbersome, which is why its not presented in its entirety here.
Each abbreviation in the INFO column should be defined in the meta section.
For example, line three from the meta region tells us that 'DP' is the raw read depth.
We can validate this by querying the meta portion.


```{r}
strwrap(grep('DP', vcf@meta, value = TRUE))
```

We see that 'DP' is defined both for the INFO column and for the gt portion of the file (FORMAT).
Returning to the INFO column, we see that the value of DP equals 197 for this variant (note the caveats in the meta line).
This tells us that this variant was sequenced 197 times over all our samples.
A number of other parameters are also included in this column which be explored by querying their definition in the meta region.
Here we just look at what is present for the first variant.


```{r, echo=TRUE}
unlist(strsplit(as.character(vcf@fix[1, 8]), split=";"))
```



## The gt region


The gt (genotype) region contains information for each variant for each sample.
The values for each variant and each sample are colon delimited.
Multiple types of data for each genotype may be stored in this manner.
The format of the data is specified by column nine, the FORMAT column.
Here we see that we have information for GT, AD, DP, GQ and PL.
The definition of these acronyms can be referenced by querying the the meta region, as demonstrated previously.
Every variant does not necessarily have the same information (e.g., SNPs and indels may be handled differently), so the rows are best treated separately.



```{r, echo=FALSE}
vcf@gt[1:6, 1:8]
```




## vcfR


Using the R package vcfR, we can read VCF format files into memory using the function `read.vcfR()`.
Once in memory we can use the `head()` method to summarize the information in the three VCF regions.



```{r}
head(vcf)
```


We now have a summary of our VCF file which we can use to help understand what forms of information are contained within it.
This information can be further explored with plotting functions and used to filter the VCF file for high quality variants.
These topics are discussed in other vignettes.



